Change history and System Architectures (#997)
open-contracting/standard: docs/getting_started/change_history.md

# Change History

> The tracking of changes in a contracting process can be important for a publisher’s open contracting goals. For example, frequent changes or amendments can signal possible corruption (red flags), or opportunities of improvement in public contracting management.

Reviewer: Seeing this in the built docs, I think we should be more direct in the introduction.

```suggestion
Tracking change over time is important for users of contracting data. For example, frequent changes or amendments can signal possible corruption (red flags), or opportunities for improvement in the management of public contracts.
```

Publishers are encouraged to publish changes to the contracting process in real time, whenever possible. OCDS supports real-time publication with the releases and records model, which describes how to publish updates to a contracting process and maintain a history of changes.

## The Releases and Records model

Releases and records are the JSON document formats supported by OCDS. These formats are not independent (records require the use of releases), but they can be seen as alternative publication formats. Publishers should decide whether to use releases, records or both to better support their goals.

Consider the timeline of a contracting process. While OCDS identifies five possible stages in the timeline (planning, tender, award, contract, implementation) there can be multiple events during each stage. The image below shows a slice of the timeline in a contracting process.

![Contracting Process](../_static/png/changehistory_process.png)

> There are major events that indicate progress from one stage to another, like the publication of an award notice, and there are smaller events that  take place within a stage, like extending the deadline for bid submission, or extending the duration of a contract. In OCDS, all changes are disclosed using **releases**. Releases are JSON documents, published each time there is any update to the contracting process.

Reviewer: For brevity we can omit the second example, since it is shown in the diagram.

```suggestion
There are major events that indicate progress from one stage to another, like the publication of an award notice, and there are smaller events that  take place within a stage, like extending the deadline for bid submission. In OCDS, all changes are disclosed using **releases**. Releases are JSON documents, published each time there is any update to the contracting process.
```

In software development terms, releases are analogous to individual Git commits on a branch, whilst a record represents both the commit history for the branch (the releases list) and HEAD for the branch (the compiled release). 

### Releases

Releases follow the OCDS Release schema, which gives them flexibility to publish small portions of data along with a few required fields. The box below shows an example.

```eval_rst
.. jsoninclude:: ../examples/tender.json
   :jsonpointer: /releases
   :expand: 
   :title: release
```

See the [Release reference](../schema/reference/) for more details on the fields available.

#### Identifiers

Each release needs to contain an ocid to identify the contracting process it relates to, and a release identifier unique in the scope of that contracting process. A release id can be built in several ways, and a publisher can use any generation strategy as long as the ids don’t clash within a contracting process.

> The Easy releases section introduces two strategies that can be used in any scenario. 

Reviewer: Reminder to add a link

See the [Release Package schema](../schema/release_package/) for details.

#### Tags

Each release also needs to provide one or more tags. Tags provide information about the event the release relates to.

For example:

* When a invitation to tender is published the ‘tender’ tag should be used and the `tender` section populated with the relevant details, however the release may also contain information in the `planning` section about the budget for the procurement.

* When an award is made the ‘award’ tag should be used and the `award` section populated with the relevant details, however the `tender` section may also be updated with the list of tenderers that submitted a bid.

The [Release tag codelist](../schema/codelists/#release-tag) contains the list of all tags provided by OCDS. Note that it is valid to use more than one tag in the same release. For example, a tenderUpdate and an award tags could be used together in a release that provides award information and also updates the number of tenderers.

#### Repeating previous information

> Releases can include new information only, or it can contain part or all unchanged data from previous releases along with the new/updated data. Both alternatives are equally valid and publishers can choose how much unchanged data to include in their releases while being careful of always including required fields. Releases with repeating information may be more appropriate for datasets with fewer releases per contracting processes. Repeating previous information also impacts on the final size of the dataset.

Reviewer: ```suggestion
Releases can include new information only, or can repeat unchanged data from previous releases along with new or updated data. Both alternatives are equally valid and publishers can choose how much unchanged data to include in their releases while being careful of always including required fields. Releases with repeating information may be more appropriate for datasets with fewer releases per contracting processes. Repeating previous information can impact on the final size of the dataset.
```

<div class="example hint" markdown=1>

<p class="first admonition-title">Example</p>

> The following example presents releases from a contracting process with minimal changes on each update. The first release presents tender data. The second introduces a new document in the tender section with the tenderUpdate tag. Note that all not-required tender fields are omitted, even the award notice present in the previous release. The third release presents award data, and the complete tender section is omitted.

Reviewer: ```suggestion
The following example presents releases from a contracting process with minimal changes on each update. The first release presents tender data. The second introduces a new document in the tender section with the tenderUpdate tag. Note that all not-required tender fields are omitted, even the tender notice present in the previous release. The third release presents award data, and the complete tender section is omitted.
```

 
 ```eval_rst
.. jsoninclude:: ../examples/minimal_updates/tender.json
   :jsonpointer: /releases/0
   :expand: tender
```
 
```eval_rst
.. jsoninclude:: ../examples/minimal_updates/tenderUpdate.json
   :jsonpointer: /releases/0
   :expand: tender
```
 
```eval_rst
.. jsoninclude:: ../examples/minimal_updates/award.json
   :jsonpointer: /releases/0
   :expand: award
```

</div>

### Records

A record should follow the structure defined in the [Records Reference](../schema/records_reference). Below is a full example.

```eval_rst
.. jsoninclude:: ../examples/merging/versioned.json
   :jsonpointer: /records/0
   :expand: 
```

#### Compiled and Versioned Releases

A **compiled release** follows the release structure and contains the latest value for all fields that have been filled in the releases of a contracting process. The example above shows how a compiled release looks like. It is strongly recommended to include a compiled release when producing records.

> A **versioned release** follows a structure similar to the release, but each final field contains an object with the history of the field. See the example above. 

Reviewer: ```suggestion
A **versioned release** follows a structure similar to the release, but in place of each field an object is provided, containing a history of changes to the field.
```

> The following animation shows how the record for a contracting process should be updated as releases are published over time.

Reviewer: Suggest moving this to the end of the page under a 'How does it all fit together?' heading

![Record updates](../_static/change_history_animation.gif)

Each time a new release is available:

* The new release is added to the releases list, either embedding it or by adding a link to it.

> * Both the compiled and versioned releases are updated with the new information from the release. Note that both should look the same either if the releases contain repeating information or not.

Reviewer: ```suggestion
* Both the compiled and versioned releases are updated with the new information from the release. Whether release repeat unchanged information from previous releases doesn't affect the resulting compiled and versioned releases.
```

<div class="example hint" markdown=1>

<p class="first admonition-title">Hint</p>

Although publishing compiled releases is not mandatory, it helps to make OCDS data more accessible to users that don’t require a detailed change history. 

Consider how to calculate the **total value of active tenders** using compiled releases:

```eval_rst
.. csv-table-no-translate::
    :header-rows: 1
    :file: ../examples/compiledreleases_compiled.csv
```

> 

Reviewer: Maybe we should make it explicit and say "Working with compiled releases, this metric can be calculated by filtering on the tender status and summing the tender value"

Compare that to how to calculate the **total value of active tenders** using releases:

```eval_rst
.. csv-table-no-translate::
    :header-rows: 1
    :file: ../examples/compiledreleases_releases.csv
```

> Whilst it’s easier to answer this simple question using the compiled releases, a full change history is required to answer other types of questions, e.g. how many tender values were amended?

Reviewer: ```suggestion
Working with individual releases, we need to check whether an active tender has later been cancelled before including it in our results.

Whilst it’s easier to answer this simple question using the compiled releases, a full change history is required to answer other types of question, such as checking how many times a tender value was amended.
```

</div>

> #### Embedding or linking Releases

Reviewer: Let's move this before the section on compiled and versioned releases, since it is a required component of a record.

Records contain the history of a contracting process as a group of releases. Releases can be embedded in the record, or a list of URLs to the releases can be provided inside it.

See the [Record reference](../schema/records_reference/) for more details.

#### Packaging

> When publishing, records should be always published wrapped in a Record Package , with a URI leading to a copy of the package and record.The following animation shows the relationship between the individual releases published during the life of a contracting process and the record for the contracting process.

Reviewer: ```suggestion
When publishing, records should be always published wrapped in a Record Package , with a URI leading to a copy of the package and record.
```

# Change History

> Tracking change over time is important for users of contracting data. Frequent changes during the procurement process can signal possible corruption. Constant, repetitive changes through processes may reveal some opportunities for improvement.

Reviewer: ```suggestion
Tracking change over time is important for users of contracting data. Frequent changes during a procurement process can signal possible corruption or opportunities for improvement.
```

![Contracting Process with releases and a record](../_static/png/changehistory_process_record.png)

Records may also contain:

* A *compiled release*, which uses the same structure as an OCDS release. It represents the latest state of each field.

* A *versioned release*, that contains a history of changes for each individual field.

These structures are designed to help users to access:

* The current state of the contracting process 
* The detailed history of changes for individual fields

In accounting terms, releases are analogous to individual transactions in an account. A record represents both the ledger of all transactions and the closing balance. The ledger is the releases list, and the closing balance is the compiled release.

In software development terms, releases are analogous to Git commits on a branch. A record represents both the commit history for the branch and the HEAD for the branch. The releases list is the commit history, and the HEAD is the compiled release.

### Releases

> Releases follow the OCDS Release schema. The schema is flexible enough to publish small portions of data, along with a few required fields. The box below shows an example.

Reviewer: ```suggestion
Releases follow the OCDS release schema. The schema covers the whole contracting process, but there are only a few required fields. The box below shows an example.
```

#### Identifiers

Each release needs to contain an [ocid](../schema/identifiers/#contracting-process-identifier-ocid) to identify the contracting process it relates to. An ocid is made by a prefix and an unique process identifier chosen by the publisher.

A release also needs a release identifier, unique in the scope of the contracting process. A release id can be built in several ways. Publishers can use any generation strategy, as long as the ids don’t clash within the same process.

The Easy releases section introduces two strategies that can be used in any scenario. 

#### Packaging

> Releases should be always wrapped in a release package to provide context to the data. The `uri` package field should contain an URI to an online copy of the same release. In a full OCDS implementation, each release should be available at its own persistent URL.

Reviewer: ```suggestion
Releases should be always wrapped in a release package to provide context to the data. The `uri` package field should contain a URI to an online copy of the same release package. In a full OCDS implementation, each release should be available at its own persistent URL.
```

    :file: ../examples/compiledreleases_releases.csv
```

If we have releases only, we need to check if any tender has been cancelled before calculating the result.

Compiled releases can be useful for many scenarios. But users may need the full history to answer other questions. For example, how many amends there are for a tender's value.

</div>

#### Embedding or linking Releases

Records contain the history of a contracting process as a group of releases. A record may have embedded releases, or a list of URLs to the releases.

See the [Record reference](../schema/records_reference/) for more details.

#### Packaging

> Records should be always published wrapped in a Record Package. The `uri` package field should lead to a copy of the same package and record.

Reviewer: ```suggestion
Records should be always published wrapped in a Record Package. The `uri` field in the package should lead to a copy of the same record package.
```
